File: desktop/lib/core/models/invoice.dart
enum InvoiceStatus {
  draft,
  pending,
  sent,
  partial,
  paid,
  overdue,
  cancelled,
}

class Invoice {
  final String id;
  final String invoiceNumber;
  final String customerId;
  final String? customerName;
  final DateTime date;
  final DateTime dueDate;
  final double subtotal;
  final double taxAmount;
  final double total;
  final double amountPaid;
  final InvoiceStatus status;
  final String currencyCode;
  final List<InvoiceLine> lines;
  final String? notes;
  final String? terms;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int? syncSequence;

  Invoice({
    required this.id,
    required this.invoiceNumber,
    required this.customerId,
    this.customerName,
    required this.date,
    required this.dueDate,
    this.subtotal = 0.0,
    this.taxAmount = 0.0,
    this.total = 0.0,
    this.amountPaid = 0.0,
    this.status = InvoiceStatus.draft,
    this.currencyCode = 'UGX',
    required this.lines,
    this.notes,
    this.terms,
    required this.createdAt,
    required this.updatedAt,
    this.syncSequence,
  });

  double get balance => total - amountPaid;
  bool get isPaid => status == InvoiceStatus.paid;
  bool get isOverdue => DateTime.now().isAfter(dueDate) && balance > 0;

  factory Invoice.fromJson(Map<String, dynamic> json) {
    return Invoice(
      id: (json['id'] ?? '').toString(),
      invoiceNumber: json['invoice_number'] as String,
      customerId: (json['customer_id'] ?? '').toString(),
      customerName: json['customer_name'] as String?,
      date: DateTime.parse(json['date'] as String),
      dueDate: DateTime.parse(json['due_date'] as String),
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0.0,
      taxAmount: (json['tax_amount'] as num?)?.toDouble() ?? 0.0,
      total: (json['total'] as num?)?.toDouble() ?? 0.0,
      amountPaid: (json['amount_paid'] as num?)?.toDouble() ?? 0.0,
      status: InvoiceStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => InvoiceStatus.draft,
      ),
      currencyCode: json['currency_code'] as String? ?? 'UGX',
      lines: (json['lines'] as List<dynamic>?)
              ?.map((l) => InvoiceLine.fromJson(l as Map<String, dynamic>))
              .toList() ??
          [],
      notes: json['notes'] as String?,
      terms: json['terms'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      syncSequence: json['sync_sequence'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoice_number': invoiceNumber,
      'customer_id': customerId,
      'customer_name': customerName,
      'date': date.toIso8601String(),
      'due_date': dueDate.toIso8601String(),
      'subtotal': subtotal,
      'tax_amount': taxAmount,
      'total': total,
      'amount_paid': amountPaid,
      'status': status.name,
      'currency_code': currencyCode,
      'lines': lines.map((l) => l.toJson()).toList(),
      'notes': notes,
      'terms': terms,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'sync_sequence': syncSequence,
    };
  }

  Invoice copyWith({
    String? id,
    String? invoiceNumber,
    String? customerId,
    String? customerName,
    DateTime? date,
    DateTime? dueDate,
    double? subtotal,
    double? taxAmount,
    double? total,
    double? amountPaid,
    InvoiceStatus? status,
    String? currencyCode,
    List<InvoiceLine>? lines,
    String? notes,
    String? terms,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? syncSequence,
  }) {
    return Invoice(
      id: id ?? this.id,
      invoiceNumber: invoiceNumber ?? this.invoiceNumber,
      customerId: customerId ?? this.customerId,
      customerName: customerName ?? this.customerName,
      date: date ?? this.date,
      dueDate: dueDate ?? this.dueDate,
      subtotal: subtotal ?? this.subtotal,
      taxAmount: taxAmount ?? this.taxAmount,
      total: total ?? this.total,
      amountPaid: amountPaid ?? this.amountPaid,
      status: status ?? this.status,
      currencyCode: currencyCode ?? this.currencyCode,
      lines: lines ?? this.lines,
      notes: notes ?? this.notes,
      terms: terms ?? this.terms,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      syncSequence: syncSequence ?? this.syncSequence,
    );
  }
}

class InvoiceLine {
  final String id;
  final String invoiceId;
  final String description;
  final double quantity;
  final double unitPrice;
  final double taxRate;
  final double taxAmount;
  final double amount;
  final String? accountId;

  InvoiceLine({
    required this.id,
    required this.invoiceId,
    required this.description,
    this.quantity = 1.0,
    this.unitPrice = 0.0,
    this.taxRate = 0.0,
    this.taxAmount = 0.0,
    this.amount = 0.0,
    this.accountId,
  });

  factory InvoiceLine.fromJson(Map<String, dynamic> json) {
    return InvoiceLine(
      id: (json['id'] ?? '').toString(),
      invoiceId: (json['invoice_id'] ?? '').toString(),
      description: json['description'] as String,
      quantity: (json['quantity'] as num?)?.toDouble() ?? 1.0,
      unitPrice: (json['unit_price'] as num?)?.toDouble() ?? 0.0,
      taxRate: (json['tax_rate'] as num?)?.toDouble() ?? 0.0,
      taxAmount: (json['tax_amount'] as num?)?.toDouble() ?? 0.0,
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      accountId: json['account_id'] == null ? null : json['account_id'].toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoice_id': invoiceId,
      'description': description,
      'quantity': quantity,
      'unit_price': unitPrice,
      'tax_rate': taxRate,
      'tax_amount': taxAmount,
      'amount': amount,
      'account_id': accountId,
    };
  }
}

File: backend/app/Models/Sales/Invoice.php
<?php

namespace App\Models\Sales;

use App\Models\Accounting\JournalEntry;
use App\Models\Company;
use App\Models\Currency;
use App\Traits\BelongsToCompany;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\MorphMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Invoice extends Model
{
    use HasFactory, SoftDeletes, BelongsToCompany;

    protected $fillable = [
        'company_id',
        'customer_id',
        'journal_entry_id',
        'client_uuid',
        'invoice_number',
        'date',
        'due_date',
        'reference',
        'notes',
        'terms',
        'currency_id',
        'exchange_rate',
        'subtotal',
        'tax_amount',
        'discount_amount',
        'total',
        'paid_amount',
        'balance',
        'status',
        'sent_at',
        'viewed_at',
        'paid_at',
    ];

    protected $casts = [
        'date' => 'date',
        'due_date' => 'date',
        'exchange_rate' => 'decimal:10',
        'subtotal' => 'decimal:4',
        'tax_amount' => 'decimal:4',
        'discount_amount' => 'decimal:4',
        'total' => 'decimal:4',
        'paid_amount' => 'decimal:4',
        'balance' => 'decimal:4',
        'sent_at' => 'datetime',
        'viewed_at' => 'datetime',
        'paid_at' => 'datetime',
    ];

    protected $attributes = [
        'exchange_rate' => 1,
        'subtotal' => 0,
        'tax_amount' => 0,
        'discount_amount' => 0,
        'total' => 0,
        'paid_amount' => 0,
        'balance' => 0,
        'status' => 'draft',
    ];

    /**
     * Status constants
     */
    const STATUS_DRAFT = 'draft';
    const STATUS_SENT = 'sent';
    const STATUS_VIEWED = 'viewed';
    const STATUS_PARTIAL = 'partial';
    const STATUS_PAID = 'paid';
    const STATUS_OVERDUE = 'overdue';
    const STATUS_CANCELLED = 'cancelled';

    /**
     * Relationships
     */
    public function company(): BelongsTo
    {
        return $this->belongsTo(Company::class);
    }

    public function customer(): BelongsTo
    {
        return $this->belongsTo(Customer::class);
    }

    public function currency(): BelongsTo
    {
        return $this->belongsTo(Currency::class);
    }

    public function journalEntry(): BelongsTo
    {
        return $this->belongsTo(JournalEntry::class);
    }

    public function lines(): HasMany
    {
        return $this->hasMany(InvoiceLine::class);
    }

    public function payments(): HasMany
    {
        return $this->hasMany(Payment::class);
    }

    public function attachments(): MorphMany
    {
        return $this->morphMany(\App\Models\Attachment::class, 'attachable');
    }

    /**
     * Calculate totals from lines
     */
    public function calculateTotals(): void
    {
        $this->subtotal = $this->lines->sum(fn($line) => $line->quantity * $line->unit_price - $line->discount_amount);
        $this->tax_amount = $this->lines->sum('tax_amount');
        $this->total = $this->subtotal + $this->tax_amount;
        $this->balance = $this->total - $this->paid_amount;
    }

    /**
     * Record payment
     */
    public function recordPayment(float $amount): void
    {
        $this->paid_amount += $amount;
        $this->balance = $this->total - $this->paid_amount;

        if ($this->balance <= 0.01) {
            $this->status = self::STATUS_PAID;
            $this->paid_at = now();
        } elseif ($this->paid_amount > 0) {
            $this->status = self::STATUS_PARTIAL;
        }

        $this->save();
    }

    /**
     * Check if overdue
     */
    public function isOverdue(): bool
    {
        return $this->due_date < now() &&
               in_array($this->status, [self::STATUS_SENT, self::STATUS_VIEWED, self::STATUS_PARTIAL]);
    }

    /**
     * Mark as sent
     */
    public function markAsSent(): void
    {
        if ($this->status === self::STATUS_DRAFT) {
            $this->status = self::STATUS_SENT;
            $this->sent_at = now();
            $this->save();
        }
    }

    /**
     * Scopes
     */
    public function scopeUnpaid($query)
    {
        return $query->whereIn('status', [self::STATUS_SENT, self::STATUS_VIEWED, self::STATUS_PARTIAL, self::STATUS_OVERDUE]);
    }

    public function scopeOverdue($query)
    {
        return $query->where('due_date', '<', now())
            ->whereIn('status', [self::STATUS_SENT, self::STATUS_VIEWED, self::STATUS_PARTIAL]);
    }

    /**
     * Boot method
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($invoice) {
            if (!$invoice->invoice_number) {
                $invoice->invoice_number = static::generateInvoiceNumber($invoice->company_id);
            }

            if (!$invoice->due_date && $invoice->customer) {
                $invoice->due_date = $invoice->date->addDays($invoice->customer->payment_terms_days);
            }
        });

        static::saved(function ($invoice) {
            // Update status to overdue if past due date
            if ($invoice->isOverdue() && $invoice->status !== self::STATUS_OVERDUE) {
                $invoice->status = self::STATUS_OVERDUE;
                $invoice->saveQuietly();
            }
        });
    }

    /**
     * Generate invoice number
     */
    private static function generateInvoiceNumber(int $companyId): string
    {
        $year = date('Y');
        $prefix = "INV-{$year}-";

        $lastInvoice = static::where('company_id', $companyId)
            ->where('invoice_number', 'like', $prefix . '%')
            ->orderByDesc('invoice_number')
            ->first();

        if ($lastInvoice) {
            $lastNumber = (int) str_replace($prefix, '', $lastInvoice->invoice_number);
            $newNumber = $lastNumber + 1;
        } else {
            $newNumber = 1;
        }

        return $prefix . str_pad($newNumber, 6, '0', STR_PAD_LEFT);
    }
}
